builtins/web/blob.h

```diff
@@ -0,0 +1,88 @@
+#ifndef BUILTINS_WEB_BLOB_H
+#define BUILTINS_WEB_BLOB_H
+
+#include "builtin.h"
+#include "extension-api.h"
+#include "js/AllocPolicy.h"
+#include "js/TypeDecls.h"
+
+namespace builtins {
+namespace web {
+namespace blob {
+
+class BlobReader {
+  std::span<const uint8_t> data_;
```

Comment:
Trying to understand the memory model here.

The blob holds onto its readers, and the stream task references the blob, so readers are guaranteed to be available while streaming, so long as there is an active stream task.

When the stream task is drained, I assume there is JS code that runs and has access to the buffer from the reader.

In this state, when there is no stream task on the task queue, and the user is directly accessing the blob buffer, what keeps the blob alive then?

Reply:
That is, I'd have expected we'd need a strong GC reference from the stream backing store to the blob?

Reply:
That's a great question. The `BlobReader` is only ever referenced in the context of `StreamTask`. What I believe keeps the blob alive is the fact that `StreamTask` maintains a reference to the blob like this:

```
    Heap<JSObject *> source_;  // -> reference to blob
```

`StreamTask` manages readers by either adding a reader state for a given stream if it doesn’t exist yet, updating or removing the reader when reading is finished. Essentially, the blob only provides a persistent reader context for the short-lived tasks.

That being said, I believe `BlobReader` and `readers` method should be removed from the public interface, as it has no meaning outside of `StreamTask` and should instead be an implementation detail of how tasks work. Otherwise, as you noticed, user could hold on to the `ReaderState` that has been invalidated.

Reply:
My concern was more the converse - if I read one chunk of a blob stream, but don't read the next one, and I'm holding a handle to the stream reader but let the blob itself be GC'd, then there's no "task" being waited on so there's no StreamTask in play - so there's nothing holding onto the blob and underlying data store, but the stream is still being held.

Reply:
In other words, the stream handler needs to hold a strong reference to the blob's data store or blob itself I think?

Reply:
Right, I kind of assumed that `NativeStream` will prevent garbage collection of a blob while it exists but I can double check that :) There are also tests for this specific scenario, both passing:

```
  "Blob.stream() garbage collection of blob shouldn't break stream consumption"
  "Blob.stream() garbage collection of stream shouldn't break stream consumption"
```

Reply:
Right, I'm just asking for clarity on the model - where is the reference from `NativeStream` to blob that ensures this?

We don't currently have proper GC test support so it's tricky to verify these behaviours. In due course we will though, so it would suck if when we start turning on agressive GC for testing a whole bunch of subsystems start breaking.

Reply:
In the `Blob::stream()` method we create a `NativeStreamSource` like this:

```cpp
  auto native_stream = streams::NativeStreamSource::create(
    cx, self, JS::UndefinedHandleValue, stream_pull, stream_cancel);
```

where `self` is a blob `JS::HandleObject`. Then `NativeStreamSource` keeps the handle as a stream owner in the slot:

```cpp
JSObject *NativeStreamSource::create(JSContext *cx, JS::HandleObject owner,
                                     JS::HandleValue startPromise,
                                     PullAlgorithmImplementation *pull,
                                     CancelAlgorithmImplementation *cancel) {
  JS::RootedObject source(cx, JS_NewObjectWithGivenProto(cx, &class_, proto_obj));
  if (!source)
    return nullptr;

  JS::SetReservedSlot(source, Slots::Owner, JS::ObjectValue(*owner));
  ...
}
```

Will this ensure that the lifetime of the blob is at least as long as that of the stream?

Reply:
Ahh okay that sounds like it resolves my question then. In addition the buffers from the stream read are by copy so there's no references to worry about there either. Thanks for talking this through.

Reply:
Thank you for the discussion here! It would be great if we could have the key properties of the memory model documented in code somewhere, as this wasn't clear to me at all, either. (See my question in `Blob::finalize` above.)

Reply:
Would it be OK if I document the model as a part of PR that switches from `std::vector` to mozilla::Vector` as this will probably change the reader quiet a bit. See my comment here: https://github.com/bytecodealliance/StarlingMonkey/pull/169#discussion_r1862493507

Reply:
Yes, absolutely!